source/utilities.cc

```diff
@@ -2535,7 +2535,67 @@ namespace aspect
       return operator_list;
     }
 
+    template <int dim>
+    GeneralizedPoint<dim>::GeneralizedPoint(Point<dim> position, const GeometryModel::Interface<dim> *geometry_model)
+    {
+      cartesian_position = position;
+      coordinate_system = geometry_model->natural_coordinate_system();
+      coordinates = geometry_model->cartesian_to_natural_coordinates(cartesian_position);
+    }
+
+    template <int dim>
+    Point<dim> &GeneralizedPoint<dim>::get_position()
+    {
+      return cartesian_position;
+    }
+
+    template <int dim>
+    std_cxx11::array<double,dim> &GeneralizedPoint<dim>::get_coordinates()
+    {
+      return coordinates;
+    }
+
+    template <int dim>
+    std_cxx11::array<double,dim-1> GeneralizedPoint<dim>::get_surface_coordinates() const
+    {
+      if (coordinate_system == Coordinates::CoordinateSystem::cartesian)
```

Comment:
Since you chain `if-else if` statements, I think it would be easier to read if you made this a `switch(coordinate_system) { ...` thingy. Then you could also add a `default: Assert (false, ExcNotImplemented());` statement at the end, where you are currently just falling off the end of the function.